src/natcap/invest/ndr/ndr_core.pyx

```diff
@@ -224,28 +202,23 @@ def ndr_eff_calculation(
                     else:
                         intermediate_retention = neighbor_effective_retention
 
-                    working_retention_eff += intermediate_retention * outflow_weight
+                    working_retention_eff += (
+                        intermediate_retention * neighbor.flow_proportion)
 
-                if outflow_weight_sum > 0:
-                    working_retention_eff /= float(outflow_weight_sum)
+                if has_outflow:
                     effective_retention_raster.set(
                         global_col, global_row, working_retention_eff)
                 else:
-                    LOGGER.error('outflow_weight_sum %s', outflow_weight_sum)
                     raise Exception("got to a cell that has no outflow!")
             # search upslope to see if we need to push a cell on the stack
-            for i in range(8):
-                neighbor_col = col_offsets[i] + global_col
-                if neighbor_col < 0 or neighbor_col >= n_cols:
-                        continue
-                neighbor_row = row_offsets[i] + global_row
-                if neighbor_row < 0 or neighbor_row >= n_rows:
-                    continue
-                neighbor_outflow_dir = inflow_offsets[i]
+            # for i in range(8):
+            for neighbor in mfd_flow_direction_raster.get_upslope_neighbors(
+                    global_col, global_row):
```

Comment:
Will this approach cause a memory leak?  Specifically: `get_upslope_neighbors` creates a new `vector` on each function call, so are we likely to get into a case where we're leaking memory on very large rasters?

From what I can tell, this is a kind of interesting case that's relatively new to the language, where the compiler should create a copy of our populated vector and return it, so the memory should be managed as stack memory in the calling function and therefore freed when the caller exits.  But this would be good to verify on a big-ish raster to verify that our memory doesn't keep steadily climbing.

Reply:
I think you are right that a copy is made. The docs [say](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html) that C++ objects can be stack allocated by using the `cdef` syntax, and then [exist only for the duration of the function](https://cython.readthedocs.io/en/latest/src/userguide/glossary.html#term-Stack-allocation). I tried to test this by comparing the memory address of the vector before and after it's returned (`print(<unsigned long>&upslope_neighbor_tuples)`) and they are indeed different addresses. But interestingly, the same address is reused in subsequent calls of `get_upslope_neighbors` - which I assume means it's being freed after each call, and then just happens to be re-allocated on the next call.

I don't have a very large MFD raster on hand to test with, but could run RouteDEM on the global DEM if you think it's worth double checking.

Reply:
I'm satisfied with this as-is!  Looking into this a little bit more, I think we will be OK.

Although the reference to the `vector` is on the stack, the vector's contents are allocated on the heap since it manages its own contents and how much memory has been allocated for it.  This means that when we return a `vector` from a function, its contents aren't necessarily erased until they are explicitly `free`d.  And then it also seems like the compiler is using some sort of [copy elision](https://en.wikipedia.org/wiki/Copy_elision) to avoid actually re-copying the memory, so my educated guess at what's causing the reference behavior you mention is that the compiler is just re-configuring the vector itself to use the same contents that are on the heap.  The fact that we're re-using the same reference makes me feel confident that the compiler is helping us out rather than creating a new object on each iteration.